src/api/mod.rs

```diff
@@ -403,6 +446,142 @@ impl<'a> ValueListBuilder<'a> {
     }
 }
 
+fn to_meta_data<'a, T>(meta_hm: T) -> Result<*mut meta_data_t, SubmitError>
+where
+    T: IntoIterator<Item = (&'a &'a str, &'a MetaValue)>,
+{
+    let meta = unsafe { meta_data_create() };
+    for (key, value) in meta_hm.into_iter() {
+        let c_key = CString::new(*key).map_err(|e| {
+            SubmitError::Field(
+                "meta key",
+                ArrayError::NullPresent(e.nul_position(), key.to_string()),
+            )
+        })?;
+        match value {
+            MetaValue::String(str) => {
+                let c_value = CString::new(str.as_str()).map_err(|e| {
+                    SubmitError::Field(
+                        "meta value",
+                        ArrayError::NullPresent(e.nul_position(), str.to_string()),
+                    )
+                })?;
+                unsafe {
+                    meta_data_add_string(meta, c_key.as_ptr(), c_value.as_ptr());
+                }
+            }
+            MetaValue::SignedInt(i) => unsafe {
+                meta_data_add_signed_int(meta, c_key.as_ptr(), *i);
+            },
+            MetaValue::UnsignedInt(u) => unsafe {
+                meta_data_add_unsigned_int(meta, c_key.as_ptr(), *u);
+            },
+            MetaValue::Double(d) => unsafe {
+                meta_data_add_double(meta, c_key.as_ptr(), *d);
+            },
+            MetaValue::Boolean(b) => unsafe {
+                meta_data_add_boolean(meta, c_key.as_ptr(), *b);
+            },
+        }
+    }
+    Ok(meta)
+}
+
+fn from_meta_data(
+    p: &str,
+    meta: *mut meta_data_t,
+) -> Result<Option<HashMap<String, MetaValue>>, ReceiveError> {
```

Comment:
My one reservation about this type that I've been waffling on over the last couple days is that the type allows the possibility of `Some` empty hashmap but the code forbids that possibility. I didn't know if you wanted to speak about why the hashmap is wrapped in an option, as to me there appears to be no semantic difference between empty meta data and no meta data. So requiring the users to differentiate these use cases seems unnecessary. I probably can be easily convinced, I just need some reassurance 😄

Reply:
You're right. there is no semantic difference for the user. I have done it this to mimic the behavior of the C-ptr which is useless.

Reply:
> It does prevent the allocation of an empty HashMap, but i don't think it worth it.

In case you'll lose sleep over this, the hashmap docs say:

> The hash map is initially created with a capacity of 0, so it will not allocate until it is first inserted into.

So if I'm reading that correctly, there shouldn't be any cost of an empty hashmap compared to `None`

Reply:
yes i've check it too :)
And when there is something, i can know call HashMap::with_capacity to have an perfectly initialized HashMap :)